File: binary search/find_smallest_letter_greater_than_target.py
class Solution:
    def nextGreatestLetter(self, letters: List[str], target: str) -> str:
        n = len(letters)
        
        l = 0
        h = n
        
        while l<h:
            mid = l + int((h - l) / 2)
            if (letters[mid] > target):
                h = mid;
            else:
                l = mid + 1; 
                
        return letters[l%n]

File: binary search/find_smallest_letter_greater_than_target.java
class Solution {
    public char nextGreatestLetter(char[] letters, char target) {
        int n = letters.length;
        
        int lo = 0, hi = n;

        
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (letters[mid] > target)     hi = mid;
            else    lo = mid + 1;                 
        }
 
        //Because lo can end up pointing to index 'n', in which case we return the first element
        return letters[lo % n];
    }
}